gr-blocks/lib/symbolinterleaver_xx_impl.cc

```diff
@@ -0,0 +1,178 @@
+#include "symbolinterleaver_xx_impl.h"
+#include <gnuradio/io_signature.h>
+#include <type_traits>
+
+namespace gr {
+namespace blocks {
+
+template <class T>
+typename symbolinterleaver_xx<T>::sptr symbolinterleaver_xx<T>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode)
+{
+    return gnuradio::make_block_sptr<symbolinterleaver_xx_impl<T>>(
+        interleaver_indices, is_packed, interleave_mode);
+}
+
+
+/*
+ * These are explicit template declarations to make sure that all these specialized make
+ * functions are present in the compiled shared library. So far this is unnecessary for
+ * other blocks. However, the reasons are yet unknown.
+ */
+template typename symbolinterleaver_xx<uint8_t>::sptr symbolinterleaver_xx<uint8_t>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode);
+
+template typename symbolinterleaver_xx<gr_complex>::sptr
+symbolinterleaver_xx<gr_complex>::make(std::vector<size_t> interleaver_indices,
+                                       bool is_packed,
+                                       bool interleave_mode);
+
+template typename symbolinterleaver_xx<float>::sptr symbolinterleaver_xx<float>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode);
+
+template typename symbolinterleaver_xx<int32_t>::sptr symbolinterleaver_xx<int32_t>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode);
+
+template typename symbolinterleaver_xx<int16_t>::sptr symbolinterleaver_xx<int16_t>::make(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode);
+
+
+template <class T>
+symbolinterleaver_xx_impl<T>::symbolinterleaver_xx_impl(
+    std::vector<size_t> interleaver_indices, bool is_packed, bool interleave_mode)
+    : gr::sync_block("symbolinterleaver",
+                     gr::io_signature::make(1, 1, sizeof(T)),
+                     gr::io_signature::make(1, 1, sizeof(T))),
+      d_is_packed(is_packed),
+      d_interleave_mode(interleave_mode),
+      d_unpacker(std::make_unique<gr::blocks::kernel::unpack_k_bits>(8)),
+      d_packer(std::make_unique<gr::blocks::kernel::pack_k_bits>(8)),
+      d_interleaver(std::make_unique<symbol_interleaving>(interleaver_indices))
+{
+    if (not std::is_same<T, uint8_t>::value) {
+        is_packed = false;
+        d_is_packed = false;
+    }
+    if (is_packed and d_interleaver->interleaver_length() % 8) {
+        throw std::invalid_argument(
+            fmt::format("Packed Interleaver requires 'interleaver_indices.size()' to be "
+                        "a multiple of 8 but got modulo8({}) = {}!",
+                        interleaver_indices.size(),
+                        interleaver_indices.size() % 8));
+    }
+    d_unpacked_original.resize(d_interleaver->interleaver_length());
+    d_unpacked_interleaved.resize(d_interleaver->interleaver_length());
+
+    gr::sync_block::set_output_multiple(is_packed
```

Comment:
Wouldn't this limit the potential size of the interleaver? I have had to implement interleavers exceeding 32K indices.

An unconstrained, pipelined streaming interleaver is a bit more logic, so for this case, perhaps have some check to on interleaver length to flag out-of-range values? Otherwise, I imagine the user will get some confusing message from the runtime.

Reply:
I have to check how GR handles `set_output_multiple` calls. I hope there's a check in there or some logic that increases the buffer size.
In any case, it'd be useful to communicate to a user why this block might fail.

I thought about supporting interleaving with larger indices then the output multiple. However, a general interleaver must assume that the last sample might be the first on the output and vice versa.
I assume that larger interleavers are designed to take this issue into account. If there's a hard limit, this info can be integrated into the docs.

Reply:
This is one of the places where `output_multiple` is used.

https://github.com/gnuradio/gnuradio/blob/eaeb3789401137fa3d7f3a30fdf782f07b331cb1/gnuradio-runtime/lib/block_executor.cc#L81

It throws in case that doesn't fit.

This is where buffers are allocated:
https://github.com/gnuradio/gnuradio/blob/eaeb3789401137fa3d7f3a30fdf782f07b331cb1/gnuradio-runtime/include/gnuradio/buffer.h#L55

And this is how a block allocates buffers. It checks `output_multiple`.
https://github.com/gnuradio/gnuradio/blob/eaeb3789401137fa3d7f3a30fdf782f07b331cb1/gnuradio-runtime/lib/block.cc#L448

Reply:
I added a test with
`nbits = int(2 ** 16) + int(2 ** 15)`
Thus, the interleaver handles `98304` items. Let's see how this works out. So far, I can confirm that it works on my machine.

Reply:
Cool, well, this was new to me. If nothing else, large interleaver tests can serve as bounds checks for individual installations.